### PR TITLE
Add relay protocol definition

### PR DESCRIPTION
**Pull request by AudriusButkevicius**
For proof reading and suggestions

### CODE REVIEW

File: specs/relay-v1.rst [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+
+Relay Protocol v1
+=================
+
+What is a relay?
+----------------
+
+Relay is a service which relays data between two *devices* which are not able to
+connect to each other directly otherwise. This is usually due to both devices
+being behind a NAT and neither sides not being able to open a port which would
```
Comment (phil-davis) [inline specs/relay-v1.rst]: 2 negatives are not needed here.
"and neither side being able to open a port..."
would be fine.

File: specs/relay-v1.rst [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+be directly accessible from the internet.
+
+A relay was designed to relay BEP protocol, hence the reliance on device ID's
+in the protocol spec, but at the same time it is general enough that could be
+reused by other protocols or aplications, as the data transfered between two
```
Comment (phil-davis) [inline specs/relay-v1.rst]: s/aplications/applications/
s/transfered/transferred/
http://forum.wordreference.com/threads/transfered-or-transferred.2920899/

File: specs/relay-v1.rst [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+Operation modes
+---------------
+
+Relay listens on a single TCP socket, but has two different connection modes,
+where a connection mode is a predefined set of messages which the relay and
+the device are expecting to exchange.
+
+The first mode is the `protocol` mode which allows a client to interact
+with the relay, for example join the relay, or request to connect to a device,
+given it is available on the relay. Simillarly to BEP, protocol mode requires
```
Comment (phil-davis) [inline specs/relay-v1.rst]: s/Simillarly/Similarly/

File: specs/relay-v1.rst [hunk 4]
```diff
@@ -0,0 +1,410 @@
...
+the connection is a protocol mode connection, due to 0x16 being the first byte
+in the TLS handshake, and only protocol mode connections use TLS.
+
+If the first byte is not 0x16, then we assume that the connection is a session
+mode connection.
+
+Protocol mode
+-------------
+
+Protocol mode uses TLS, and protocol name defined by the TLS heander should be
```
Comment (phil-davis) [inline specs/relay-v1.rst]: "and the protocol name"
s/heander/header/

File: specs/relay-v1.rst [hunk 5]
```diff
@@ -0,0 +1,410 @@
...
+	/                                                               /
+	+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+
+
+	struct JoinSessionRequest {
+		opaque Key<32>;
+	}
+
+: Key
+	This is a unique random session key generated by the relay server. It's is
```
Comment (phil-davis) [inline specs/relay-v1.rst]: "It's is" => "It is"

File: specs/relay-v1.rst [hunk 6]
```diff
@@ -0,0 +1,410 @@
...
+		opaque Key<32>;
+		opaque Address<32>;
+		unsigned int Port;
+		bool ServerSocket;
+	}
+
+: From
+	Device ID identifying who you will be connecting with.
+: Key
+	A unique random session key generated by the relay server. It's is used to
```
Comment (phil-davis) [inline specs/relay-v1.rst]: "It's is" => "It is"

File: specs/relay-v1.rst [hunk 7]
```diff
@@ -0,0 +1,410 @@
...
+	identify which session you are trying to connect to.
+: Address
+	An optional IP address on which the relay server is expecting you to
+	connect, in order to start a connection in session mode.
+	Empty/all zero IP should be replaced with the  relays public IP address that
```
Comment (phil-davis) [inline specs/relay-v1.rst]: "the   relays" => "the relay's"

File: specs/relay-v1.rst [hunk 8]
```diff
@@ -0,0 +1,410 @@
...
+	the server side or the client side, this boolean indicates which side of the
+	connection this client should assume it's getting. The value is inverted in
+	the invitation which is sent to the other device, so that there is always
+	one client socket, and one server socket.
+
+How syncthing uses relays, and general security
+-----------------------------------------------
+
+In the case of Syncthing and BEP, when two devices connect via relay, they
+start their standard TLS connection encapsulated within the relays plain-text
```
Comment (phil-davis) [inline specs/relay-v1.rst]: relay's

**Comment by phil-davis** [comment]
I would be nice to also say what BEP stands for.

**Comment by calmh** [comment]
LGTM ship it

**Pull request merged**